pkg/cmd/pulumi/metadata/metadata.go

```diff
@@ -220,6 +224,60 @@ func addGitMetadata(projectRoot string, m *backend.UpdateMetadata) error {
 	return allErrors.ErrorOrNil()
 }
 
+// addGitMetadataFromEnvironment retrieves Git-related metadata from environment variables in the case that a Git
+// repository is not present and adds any available information to the given metadata object.
+func addGitMetadataFromEnvironment(m *backend.UpdateMetadata) {
+	if owner := os.Getenv("PULUMI_VCS_REPO_OWNER"); owner != "" {
+		m.Environment[backend.VCSRepoOwner] = owner
+	}
+	if repo := os.Getenv("PULUMI_VCS_REPO_NAME"); repo != "" {
+		m.Environment[backend.VCSRepoName] = repo
+	}
+	if kind := os.Getenv("PULUMI_VCS_REPO_KIND"); kind != "" {
+		m.Environment[backend.VCSRepoKind] = kind
+	}
+	if root := os.Getenv("PULUMI_VCS_REPO_ROOT"); root != "" {
+		m.Environment[backend.VCSRepoRoot] = root
+	}
+
+	// As in addGitCommitMetadata, which we'd call if we had a Git repository, we'll use any information provided by the
+	// CI system (or appropriate environment variables) to populate remaining Git-related fields that we don't have
+	// explicit values for.
+	vars := ciutil.DetectVars()
+
+	message := os.Getenv("PULUMI_GIT_COMMIT_MESSAGE")
+	if message == "" {
+		message = vars.CommitMessage
+	}
+	if message != "" {
+		m.Message = gitCommitTitle(message)
+	}
+
+	headName := os.Getenv("PULUMI_GIT_HEAD_NAME")
+	if headName == "" {
+		headName = vars.BranchName
+	}
+	if headName != "" && headName != "HEAD" {
```

Comment:
why `headName != HEAD` here?  If the user explicitly sets the variable, shouldn't we still respect it?  Not that I think it would make a lot of sense for me, but still.

Reply:
Good call -- have made the change and left a comment explaining the difference between this and auto-detect. Thanks!